types/index.d.ts

```diff
@@ -156,6 +156,131 @@ declare namespace math {
     isHexDigit(c: string): boolean;
   }
 
+  interface AccessorNode extends MathNode {
+    object: MathNode;
+    index: IndexNode;
+    name: string;
+  }
+  interface AccessorNodeCtor {
+    new(object: MathNode, index: IndexNode): AccessorNode;
+  }
+
+  interface ArrayNode extends MathNode {
+    items: MathNode[];
+  }
+  interface ArrayNodeCtor {
+    new(items: MathNode[]): ArrayNode;
+  }
+
+  interface AssignmentNode {
+    object: SymbolNode | AccessorNode;
+    index: IndexNode | null;
+    value: MathNode;
+    name: string;
+  }
+  interface AssignmentNodeCtor {
+    new(object: SymbolNode, value: MathNode): AssignmentNode;
+    new(object: SymbolNode | AccessorNode, index: IndexNode, value: MathNode): AssignmentNode;
+  }
+
+  interface BlockNode extends MathNode {
+    blocks: Array<{node: MathNode, visible: boolean}>;
+  }
+  interface BlockNodeCtor {
+    new(arr: Array<{node: MathNode} | {node: MathNode, visible: boolean}>): BlockNode;
+  }
+
+  interface ConditionalNode extends MathNode {
+    condition: MathNode;
+    trueExpr: MathNode;
+    falseExpr: MathNode;
+  }
+  interface ConditionalNodeCtor {
+    new(condition: MathNode, trueExpr: MathNode, falseExpr: MathNode): ConditionalNode;
+  }
+
+  interface ConstantNodeCtor {
+    new(constant: number): MathNode;
+  }
+
+  interface FunctionAssignmentNode extends MathNode {
+    name: string;
+    params: string[];
+    expr: MathNode;
+  }
+  interface FunctionAssignmentNodeCtor {
+    new(name: string, params: string[], expr: MathNode): FunctionAssignmentNode;
+  }
+
+  interface FunctionNode extends MathNode {
+    //TODO Proper type for fn
+    // fn: MathNode | string;
```

Comment:
If I'm not mistaken, `functionNode.fn` is always a `MathNode`, as can be seen [here](https://github.com/josdejong/mathjs/blob/feature/async_support/src/expression/node/FunctionNode.js#L31).

Reply:
Ah, I see, there's a problem with overwriting `MathNode`. I think that the proper way to resolve this would be to have a `MathNodeCommon` interface which implements all the common properties of all nodes, make all nodes inherit from `MathNodeCommon` instead of `MathNode` and then declare
```typescript
type MathNode = AccessorNode | ArrayNode | AssignmentNode | ...`
```